File: examples/Examples.Language.Vb/TVShow.vb
Namespace Objects

    ' The VB half of the objects section.  Note that these are Property and not
    ' ReadOnly Property: a `With {}` initializer assigns to them, and a ReadOnly
    ' auto-property has no setter for it to assign through.
    Public Class TVShow
        Public Sub New()
            ' constructor
        End Sub

        Private _showName As String

        ' Public properties can be accessed from any function inside the
        ' class as well as other classes
        Public Property ShowName() As String
            Get
                ' Inside the get part the private variable is returned.
                ' You can do anything you want here such as data validation
                ' before returning the data if you need or want.
                Return _showName
            End Get
            Set(ByVal value As String)
                ' Inside the set part the private variable is set.
                ' You can do anything you want here such as data validation
                ' before the data is set.
                If value.Trim = "" Then
                    Throw New Exception("ShowName cannot be empty")
                End If
                _showName = value
            End Set
        End Property

        ' The above property is long form.  A shorter form can be done as seen below
        Public Property ShowLength As Integer
        Public Property Summary As String
        Public Property Rating As Decimal
        Public Property Episode As String
    End Class

End Namespace

File: examples/Examples.Language/Objects/TVShow.cs
namespace Examples.Language.Objects;

// The chapter's running example.  A long-form property with validation, and the
// short form for the rest.
public class TVShow
{
    public TVShow()
    {
    }

    // Assigned so the compiler knows it is never null before ShowName's setter
    // runs.  With nullable reference types on, a non-nullable field has to hold
    // something by the time the constructor exits.
    private string _showName = string.Empty;

    // Public properties can be accessed from any function inside the
    // class as well as other classes.  required means an object initializer has
    // to set it, so the type cannot be constructed half built.
    public required string ShowName
    {
        get
        {
            // Inside the get part the private variable is returned.
            // You can do anything you want here such as data validation
            // before returning the data if you need or want.
            return _showName;
        }
        set
        {
            // Inside the set part the private variable is set.
            // You can do anything you want here such as data validation
            // before the data is set.
            if (value.Trim() == "")
                throw new Exception("ShowName cannot be empty");
            _showName = value;
        }
    }

    // The above property is long form.  A shorter form can be done as seen below
    public int ShowLength { get; init; }
    public required string Summary { get; init; }
    public decimal Rating { get; init; }
    public required string Episode { get; init; }
}
